scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/synchronization/AOSynchronization.java

```diff
@@ -0,0 +1,862 @@
+                                                                                                                                                                    BiFunction.class)));
+            } else {
+                throw new IllegalStateException("Predicate " + predicate + " is not met");
+            }
+            commitIfNeeded(channel);
+            logWithContext(taskid,
+                           channel,
+                           "Run waitUntil" + ON_KEY + QUOTE + key + QUOTE + RESULT_IS + answer,
+                           Level.DEBUG);
+        } catch (InvalidChannelException | CompilationException | IOException e) {
+            throw e;
+        } catch (Exception e) {
+            throw new ClosureEvaluationException(EXCEPTION_WHEN_EVALUATING_CLOSURE + StackTraceUtil.getStackTrace(e));
+        }
+        return answer;
+    }
+
+    private void logWithContext(TaskId taskId, String channel, String message, Level level)
+            throws InvalidChannelException {
+        String channelHeader = "[Synchronization]" + (channel != null ? "[Channel=" + channel + "] " : "");
+        switch (level.toInt()) {
+            case Level.TRACE_INT:
+                if (logger.isTraceEnabled()) {
+                    TaskLogger.getInstance().info(taskId, channelHeader + message);
+                }
+                break;
+            case Level.DEBUG_INT:
+                if (logger.isDebugEnabled()) {
+                    TaskLogger.getInstance().info(taskId, channelHeader + message);
+                }
+                break;
+            case Level.INFO_INT:
+                if (logger.isInfoEnabled()) {
+                    TaskLogger.getInstance().info(taskId, channelHeader + message);
+                }
+                break;
+            case Level.WARN_INT:
+                TaskLogger.getInstance().warn(taskId, channelHeader + message);
+                break;
+            case Level.ERROR_INT:
+                TaskLogger.getInstance().error(taskId, channelHeader + message);
+                break;
+            default:
+                // no action
+        }
+        if (logger.isTraceEnabled() && channel != null) {
+            TaskLogger.getInstance().info(taskId, channelHeader + "New channel content : " + getChannel(channel));
+        }
+    }
+
+    @Override
+    public void runActivity(Body body) {
+        Service service = new Service(body);
+        while (body.isActive()) {
+            try {
+                Request request = service.blockingRemoveOldest();
+                if (request == null) {
+                    return;
+                }
+                if (request.getMethodName().startsWith("wait") && !testWaitFunction(service, request)) {
+                    waitUntilQueue.add(request);
+                } else {
+                    service.serve(request);
+                    unblockWaitMethods(service);
```

Comment:
I think the call to `unblockWaitMethods(service)` should be moved outside of the if/else statement. Here is why:
suppose the request predicate is already satisfied. We would put the request in the queue and loop again. Then we serve another request that can be served immediately, so we execute it. Then we call `unblockWaitMethods` and we finally execute the request that was put in the `waitUntilQueue`. In the end we swapped the order of execution of the two requests for no reason because the first one could have been executed immediately.

Reply:
I don't understand: if the request predicate is already satisfied, then the request is NOT added in the queue and immediately served